### PR TITLE
Remove partial functions from the codebase

### PR DESCRIPTION
**Pull request by scruffystuffs**
# Overview

Remove partial functions from the codebase.  We only really care about partial functions in the code we ship, so test- and compile-time code are ignored.

### CODE REVIEW

File: src/App/Fossa/BinaryDeps/Jar.hs [hunk 1]
```diff
@@ -56,7 +56,7 @@ parseMetaInfManifest :: Text -> Map Text Text
 parseMetaInfManifest t = Map.fromList . map strip' . filter' $ map (Text.breakOn ":") (Text.lines t)
   where
     null' (a, b) = any Text.null [a, b]
-    strip' (a, b) = (Text.strip a, Text.strip $ Text.tail b)
+    strip' (a, b) = (Text.strip a, Text.strip $ Text.drop 1 b)
```
Comment (scruffystuffs) [inline src/App/Fossa/BinaryDeps/Jar.hs:59]: cc @kitified

File: src/App/Fossa/VPS/Scan/RunWiggins.hs [hunk 2]
```diff
@@ -148,7 +149,9 @@ optExplodeText flag (a : as) = [flag, a] ++ optExplodeText flag as
 -- Path Rel Dir renders with a trailing /, but wiggins needs it to not have that trailing slash when used as an exclude or include-only filter.
 -- Strip the / postfix from the returned value.
 optPathAsFilter :: Path Rel Dir -> Text
-optPathAsFilter p = Text.init (toText p)
+optPathAsFilter p = fromMaybe t $ Text.stripSuffix "/" t
+  where
+    t = toText p
```
Comment (scruffystuffs) [inline src/App/Fossa/VPS/Scan/RunWiggins.hs:154]: cc @kitified

File: src/Data/List/Extra.hs [hunk 3]
```diff
@@ -4,6 +4,6 @@ module Data.List.Extra ((!?)) where
 -- if the index doesn't exist, return Nothing, otherwise
 -- return Just (value)
 (!?) :: [a] -> Int -> Maybe a
-xs !? ix
-  | length xs <= ix = Nothing
-  | otherwise = Just (xs !! ix)
+xs !? ix = case drop ix xs of
+  [] -> Nothing
+  a : _ -> Just a
```
Comment (scruffystuffs) [inline src/Data/List/Extra.hs:9]: cc @cnr

File: src/Strategy/Carthage.hs [hunk 4]
```diff
@@ -136,7 +137,7 @@ entryToCheckoutName entry =
   case resolvedType entry of
     GitEntry -> resolvedName entry
     -- this is safe because Text.splitOn always returns a non-empty list
-    GithubType -> last . Text.splitOn "/" $ resolvedName entry
+    GithubType -> snd . TextExt.splitOnceOnEnd "/" $ resolvedName entry
```
Comment (scruffystuffs) [inline src/Strategy/Carthage.hs]: cc @cnr

File: src/Strategy/Go/Types.hs [hunk 5]
```diff
@@ -71,8 +72,8 @@ golangPackageToDependency pkg = foldr applyLabel start
 -- TODO: In `go.mod`, we handle this in the parser. Do we even need to handle
 -- this format in other Go analyzers? Can we just remove this code?
 fixVersion :: Text -> Text
-fixVersion = last . Text.splitOn "-" . Text.replace "+incompatible" ""
+fixVersion = snd . Text.splitOnceOnEnd "-" . Text.replace "+incompatible" ""
```
Comment (scruffystuffs) [inline src/Strategy/Go/Types.hs:75]: cc @cnr

File: src/Strategy/Go/Types.hs [hunk 6]
```diff
@@ -71,8 +72,8 @@ golangPackageToDependency pkg = foldr applyLabel start
...
 
 -- replace "github.com/A/B/vendor/github.com/X/Y" with "github.com/X/Y"
 unvendor :: Text -> Text
-unvendor = last . Text.splitOn "/vendor/"
+unvendor = snd . Text.splitOnceOnEnd "/vendor/"
```
Comment (scruffystuffs) [inline src/Strategy/Go/Types.hs:79]: cc @cnr

**Review by scruffystuffs — COMMENTED** [review]
Left tags for the original implementors to review.

File: .hlint.yaml [hunk 7]
```diff
@@ -20,7 +20,22 @@
...
+  - {name: Data.Bytestring.tail, within: []}
+  - {name: Data.Bytestring.init, within: []}
+  - {name: Data.Bytestring.last, within: []}
+  - {name: Data.Text.head, within: []}
+  - {name: Data.Text.tail, within: []}
+  - {name: Data.Text.init, within: []}
+  - {name: Data.Text.last, within: []}
+  - {name: "Prelude.!!", within: [Elixir.MixTreeSpec]}
+  - {name: "Data.Map.!", within: []}
+  - {name: fromJust, within: [App.DocsSpec]}
```
Comment (scruffystuffs) [inline .hlint.yaml:41]: Are there more we should add?

File: src/Strategy/Carthage.hs [hunk 8]
```diff
@@ -136,7 +136,7 @@ entryToCheckoutName entry =
   case resolvedType entry of
     GitEntry -> resolvedName entry
     -- this is safe because Text.splitOn always returns a non-empty list
-    GithubType -> last . Text.splitOn "/" $ resolvedName entry
+    GithubType -> snd . Text.splitOnceOnEnd "/" $ resolvedName entry
```
Comment (scruffystuffs) [inline src/Strategy/Carthage.hs:139]: cc @cnr

**Review by kitified — APPROVED** [review]
👍 on the ones tagged to me!

File: src/Strategy/Carthage.hs (hunk 8 above)
Comment (cnr) [reply src/Strategy/Carthage.hs:139]: should kill the "this is safe because..." comments on each of these `Text.splitOn` changes

**Review by cnr — APPROVED** [review]

**Comment by meghfossa** [comment]
Merged as wes is on vacation - wanted to get this in prior to repo switchover.

**Pull request merged**